### PR TITLE
[ffigen] ffigen only regenerates if input mtimes  > output mtimes (#874)

### PR DESCRIPTION
**Pull request by Abdk4Moura**
Before this, ffigen generates and overwrites any previously generated dart bindings. Now, it checks if any of the input files has been updated (including the config file itself) and only regenerates if the input files have indeed been regenerated.

### CODE REVIEW

File: pkgs/ffigen/CHANGELOG.md [hunk 1]
```diff
@@ -6,7 +11,7 @@
 
 ## 11.0.0
 
-- Any compiler errors/warnings in source header files will now result in
+- Any compilier errors/warnings in source header files will now result in
```
Comment (dcharkes) [inline pkgs/ffigen/CHANGELOG.md:14]: typo

**Review by dcharkes — CHANGES_REQUESTED** [review]
@Abdk4Moura Thanks!

Please add tests. The tests should also cover situations where directories and glob patterns are used instead of only specific files.

File: pkgs/ffigen/CHANGELOG.md (hunk 1 above)
Comment (Abdk4Moura) [reply pkgs/ffigen/CHANGELOG.md]: Thanks! just corrected the typo

**Comment by Abdk4Moura** [comment]
I have added tests for the `_inputHasBeenUpdated` function.

**Comment by Abdk4Moura** [comment]
But does it make a difference with globs and normal file names. Globs, as they seem to me, must have already been resolved by either the `Config` config or `Library` library objects.

Is a test with respect to those still relevant?

**Comment by dcharkes** [comment]
Sorry for the delay, I have too many tasks at the moment.

Any chance @liamappelbe and @mannprerak2 can review this instead?

**Comment by mannprerak2** [comment]
No problem @dcharkes :D

@Abdk4Moura this only considers the last modified times for entrypoints headers declared in config. But it's possible some other imported header file might have changed instead.

Any transitive dependencies could be changed and will be missed by this.

Some more edge cases to consider -
1. ffigen version might be updated in pubspec.yaml
2. Any updates to compiler-opts specified via the environment itself.

**Comment by liamappelbe** [comment]
I'm hesitant to add this at all. In my experience with rolling my own build system, it's actually surprisingly difficult to catch all the edge cases where deps have been modified. It would be really frustrating for our users, and difficult for them to diagnose, if we missed one of those edge cases, and ffigen was just not running.

My preference would be to let a higher level of build system handle checking the file times, like Bazel or [package:build](https://pub.dev/packages/build) or [native assets](https://pub.dev/packages/native_assets_cli).

If people really want this, then we at least need a command line flag and/or a config option to force rebuild.

**Comment by Abdk4Moura** [comment]
I was thinking of that too; a "--force" flag

**Comment by dcharkes** [comment]
Thanks @mannprerak2 and @liamappelbe!

Based on your comments, some high level observations.

> My preference would be to let a higher level of build system handle checking the file times, like Bazel or [package:build](https://pub.dev/packages/build) or [native assets](https://pub.dev/packages/native_assets_cli).

👍  Great thinking about how this integrates in the larger eco system.

I think the FFIgen (and JNIgen) should be able to report the list of dependencies (https://github.com/dart-lang/native/issues/1008). We don't want every user of FFIgen (and JNIgen) to manually come up with a list of dependencies.

The logic of checking the files timestamps could maybe live in a shared package that then FFIgen, JNIgen, and native_assets_builder can reuse.

One question to ask is what abstraction to use for listing dependencies:

1. File paths
2. Directory paths (which would mean all files transitively in that directory, and it also means that if someone adds a file in such directory it's added).
3. Glob patterns? (which would mean that any files added to the file system matching such pattern would trigger a rebuild). -> But not every build system wrapping FFIgen/JNIgen might be able consume glob patterns (native assets, package:build, ...)

**Comment by dcharkes** [comment]
I have distilled some of the learnings from the PR into https://github.com/dart-lang/native/issues/1101.

Thanks @Abdk4Moura for the suggestion and the exploration!

I don't believe we can land this PR in it's current state, I'll close it.

**Pull request closed without merging**